### DISCUSSION TITLE
Interoperability of fips approved providers

### DISCUSSION THREAD

**Discussion by afshinpir** [openssl/openssl#24572] [Q&A] [answered]
Hi all

Let's assume that I have a fips-approved provider that implements a hardware-based AES. Now I want to use OpenSSL's CMAC with this provider. If I create CMAC using default provider and pass my providers query string via init parameters, everything works fine. But if I create CMAC using fips provider, I will not be able to connect it to a 3rd party provider. I wonder if there is a interoperability method between fips provider and any 3rd party fips-approved providers or not.
As a second question, am I allowed to add `fips=yes` to my provider's query string? And if I do, will my provider be usable in fips context?

Best Regards

**Comment by nhorman** [comment, marked as answer]
In answer to your first question, the answer I believe is no.  FIPS approval of a given fips module is predicated on the definition of a module boundary, which for the openssl FIPS certificate is the fips provider itself.  Its not valid to export the work of a given fips module to a different implementation (in this case your provider and its associated offload hardware).  The fact that your provider is also fips validated isn't relevant, as there is no certificate that defines a fips module that encompasses both.  It would certainly be possible to create such a certificate by going through the appropriate process with your lab, but you need to bring that up with them.

As to the second question, that is I believe up to you.  If you have a fips provider that you have validated with a lab, then it would be reasonable to add the element to your provider string.  The impact to your fips certificate is a question you need to take up with your lab however

**Reply by paulidale** [reply]
OpenSSL's FIPS provider pulls all required algorithms from inside its hidden internal library context.  So CMAC from the FIPS provider will only ever use the FIPS provider's internal AES implementation.  This was by design to avoid accidentally ending up in a non-approved mode.  It does have the unfortunate short coming that your validated provider's implementation won't be used.

@nhorman isn't quite correct saying that FIPS boundaries cannot be crossed.  There are provisions in the standards to do this, so it would theoretically be possible to have a FIPS module with CMAC using AES from a separate independent FIPS module.  Practically, however, this is quite problematic: any data transferred between the two boundaries has to use a _secure channel_ which mandates an integrity check be included.  Just having an integrity check isn't actually preventing tampering, for that something more akin to a full TLS null cipher connection is required (which is crazy).

Something we might want to consider at some point in the, likely distant, future would be a capability to add additional providers to the FIPS provider's internal library context.  Because the provider APIs are stable, it would theoretically be possible and it would give you the functionality you require.  There is a lot to think through before attempting such.
